src/lib.rs

```diff
@@ -63,3 +70,114 @@ pub fn downsample(src: &Image, target_width: u32, target_height: u32) -> Vec<u8>
 
     output
 }
+
+#[derive(Debug, Clone)]
+struct CachedWeight {
+    pub start: u32,
+    pub coefficients: Rc<Vec<f32>>,
+}
+
+fn calculate_weights(src: u32, target: u32) -> Vec<CachedWeight> {
+    assert!(
+        src > target,
+        "Trying to use downsampler to upsample or perform an operation which will cause no changes"
+    );
+
+    let mut variables = vec![WeightVariables::default(); target as usize];
+
+    unsafe {
+        ispc::downsample_ispc::calculate_weight_variables(src, target, variables.as_mut_ptr());
+    };
+
+    let image_scale = src as f32 / target as f32;
+
+    let mut res = Vec::with_capacity(target as usize);
+
+    let mut reuse_heap = HashMap::<_, Rc<Vec<f32>>>::with_capacity(target as usize / 2);
```

Comment:
It this something we should `static`ally cache (https://crates.io/crates/once_cell) so that subsequent `downsample()` calls benefit from it? And/or put it in the public signature behind an opaque function so that the caller is in control over how to share - _and when to destroy to free up memory_ - the cache?